internal/dataplane/kongstate/kongstate.go

```diff
@@ -690,3 +683,303 @@ func maybeLogKongIngressDeprecationError(logger logr.Logger, services []*corev1.
+			k8sKongCustomEntity: entity,
+		})
+	}
+
+	ks.sortCustomEntities()
+}
+
+// fetchEntitySchema fetches schema of an entity by its type and stores the schema in its custom entity collection
+// as a cache to avoid excessive calling of Kong admin APIs.
+func (ks *KongState) fetchEntitySchema(schemaGetter SchemaGetter, entityType string) (EntitySchema, error) {
+	collection, ok := ks.CustomEntities[entityType]
+	if ok {
+		return collection.Schema, nil
+	}
+	// Use `context.Background()` here because `BuildKongConfig` does not provide a context.
+	schema, err := schemaGetter.Get(context.Background(), entityType)
+	if err != nil {
+		return EntitySchema{}, err
+	}
+	return ExtractEntityFieldDefinitions(schema), nil
+}
+
+// fillCustomEntityForeignFields fills the "foreign" fields of a custom Kong entity
+// if it refers to a service/route/consumer.
+// Because Kong gateway requires the "foreign" fields to use IDs of the referred entity,
+// it fills ID of the referred entity and fill the ID of the entity to the field.
+// So the function has the side effect that the referred entity will have a generated fixed ID.
+// It does not support fields referring to other types of entities.
+func (ks *KongState) fillCustomEntityForeignFields(
+	logger logr.Logger,
+	k8sEntity *kongv1alpha1.KongCustomEntity,
+	schema EntitySchema,
+	parsedEntity map[string]any,
+	pluginRelEntities PluginRelatedEntitiesRefs,
+	workspace string,
+) {
+	logger = logger.WithValues("entity_namespace", k8sEntity.Namespace, "entity_name", k8sEntity.Name)
+	// Find referred entity via the plugin in its spec.parentRef.
+	// Then we can fetch the referred service/route/consumer from the reference relations of the plugin.
+	parentRef := k8sEntity.Spec.ParentRef
+	var namespace string
+	// Abort if the parentRef is empty or does not refer to a plugin.
+	if parentRef == nil ||
+		(parentRef.Group == nil || *parentRef.Group != kongv1alpha1.GroupVersion.Group) {
+		return
+	}
+	if parentRef.Kind == nil || (*parentRef.Kind != "KongPlugin" && *parentRef.Kind != "KongClusterPlugin") {
+		return
+	}
+	// Extract the plugin key to get the plugin relations.
+	if parentRef.Namespace == nil || *parentRef.Namespace == "" {
+		namespace = k8sEntity.Namespace
+	} else {
+		namespace = *parentRef.Namespace
+	}
+	pluginKey := namespace + ":" + parentRef.Name
+	// Get the relations with other entities of the plugin.
+	rels, ok := pluginRelEntities.RelatedEntities[pluginKey]
+	if !ok {
+		return
+	}
+	logger.V(util.DebugLevel).Info("fetch references via plugin", "plugin_key", pluginKey)
+
+	// Traverse through the fields of the entity and fill the "foreign" fields with IDs of referring entities.
+	// Note: this procedure will make referred services'/routes'/consumers' ID to be filled. Should we set `FillIDs` feature gate as a prerequisite?
+	for fieldName, field := range schema.Fields {
+		if field.Type != EntityFieldTypeForeign {
+			continue
+		}
+		switch field.Reference {
+		case "services":
```

Comment:
Do we have `const`s for `case` values used in here"? (`services`, `routes` etc.)

Reply:
Added the `const`s from definitions of `kong.EntityType`s in `go-kong`.